runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -211,6 +211,372 @@ doubleMaxMinHelper(TR::Node *node, TR::CodeGenerator *cg, bool isMaxOp)
+ *
+ * <\verbatim>
+ *
+ *
+ * \param node the intrinsic function call node
+ * \param cg the code generator
+ * \param elementSize element size in bytes. This is 1 for compact strings of Latin1 letters, and it's 2 for non-compact strings.
+ *
+*/
+TR::Register*
+inlineVectorizedStringIndexOf(TR::Node* node, TR::CodeGenerator* cg, uint32_t elementSize)
+   {
+   // element size is 2-byte if both s1 and s2 are decompressed strings and it is 12 otherwise.
+   // 0 for byte, 1 for halfword
+   uint32_t elementSizeMask = (elementSize == 1) ? 0 : 1;
+   bool isUTF16 = (elementSize == 2);
+
+   const int8_t vectorSize = cg->machine()->getVRFSize();
+   const bool is64Bit = TR::Compiler->target.is64Bit();
+   uintptrj_t headerSize = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+   TR::Compilation* comp = cg->comp();
+
+   if (comp->getOption(TR_TraceCG))
+      traceMsg(comp, "inlineVectorizedStringIndexOf element size %d\n", elementSize);
+
+   // Get call parameters
+   // s1Value and s2Value are byte arrays
+   // This evaluator function handles different indexOf() instrinsics, some of which are static calls without a receiver. Hence, the need for static call check.
+   bool isStaticCall = node->getSymbolReference()->getSymbol()
+           && node->getSymbolReference()->getSymbol()->castToMethodSymbol()
+           && node->getSymbolReference()->getSymbol()->castToMethodSymbol()->isStatic();
+   uint8_t firstCallArgIdx = isStaticCall ? 0 : 1;
+
+   TR::Register* r_s1Value   = cg->evaluate          (node->getChild(firstCallArgIdx));
+   TR::Register* r_s1Len     = cg->gprClobberEvaluate(node->getChild(firstCallArgIdx+1));
+   TR::Register* r_s2Value   = cg->evaluate          (node->getChild(firstCallArgIdx+2));
+   TR::Register* r_s2Len     = cg->gprClobberEvaluate(node->getChild(firstCallArgIdx+3));
+   TR::Register* r_fromIndex = cg->gprClobberEvaluate(node->getChild(firstCallArgIdx+4));
+
+   // Registers
+   // r_tmp2 is used to eliminate data dependency in the VL-VL-VCEQ loop
+   TR::Register* r_resultReg       = cg->allocateRegister();
+   TR::Register* r_maxIndex        = cg->allocateRegister();
+   TR::Register* r_s1VecStartIndex = cg->allocateRegister();
+   TR::Register* r_s2VecStartIndex = cg->allocateRegister();
+   TR::Register* r_loadLen         = cg->allocateRegister();
+   TR::Register* r_tmp             = cg->allocateRegister();
+   TR::Register* r_tmp2            = cg->allocateRegister();
+   TR::Register* r_1stCharIndex    = cg->allocateRegister();
+
+   TR::Register* v_s1Partial       = cg->allocateRegister(TR_VRF);
+   TR::Register* v_s2Partial       = cg->allocateRegister(TR_VRF);
+   TR::Register* v_s2Char1Rep      = cg->allocateRegister(TR_VRF);
+   TR::Register* v_tmp             = cg->allocateRegister(TR_VRF);
+
+   // register dependencies
+   TR::RegisterDependencyConditions * regDeps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 17, cg);
```

Comment:
This is a scary number of dependencies. I count 13 GPRs which is almost more than we have available. The register pressure here is going to be huge causing a ton of spilling of live values. Is there a way to reduce the number of registers you need for this evaluator?

Reply:
Yes. I'm worried about this, too. I can probably eliminate one GPR and see if there are other opportunities.

The problem here is that that we have 5 parameters, and two strings to point into. The number of variables here is naturally large.

Reply:
Reduced the number of GPRs from 13 to 10.